src/target/riscv/riscv-013.c

```diff
@@ -3396,329 +3396,567 @@ static int write_memory_abstract(struct target *target, target_addr_t address,
+{
+	RISCV013_INFO(info);
+	/* s0 holds the next address to read from
+	 * s1 holds the next data value read
+	 * a0 is a counter in case increment is 0
+	 */
+	if (register_write_direct(target, GDB_REGNO_S0, address + index * increment)
+			!= ERROR_OK)
+		return ERROR_FAIL;
 
-				dmi_write(target, DM_ABSTRACTAUTO, 0);
+	if (/*is_repeated_read*/ increment == 0 &&
+			register_write_direct(target, GDB_REGNO_A0, index) != ERROR_OK)
+		return ERROR_FAIL;
 
-				uint32_t dmi_data0, dmi_data1 = 0;
-				/* This is definitely a good version of the value that we
-				 * attempted to read when we discovered that the target was
-				 * busy. */
-				if (dmi_read(target, &dmi_data0, DM_DATA0) != ERROR_OK) {
-					riscv_batch_free(batch);
-					goto error;
-				}
-				if (size > 4 && dmi_read(target, &dmi_data1, DM_DATA1) != ERROR_OK) {
-					riscv_batch_free(batch);
-					goto error;
-				}
+	const uint32_t startup_command =
+		read_memory_progbuf_inner_startup_command(target);
+	if (execute_abstract_command(target, startup_command) != ERROR_OK)
+		return ERROR_FAIL;
 
-				/* See how far we got, clobbering dmi_data0. */
-				if (increment == 0) {
-					uint64_t counter;
-					result = register_read_direct(target, &counter, GDB_REGNO_A0);
-					next_index = counter;
-				} else {
-					uint64_t next_read_addr;
-					result = register_read_direct(target, &next_read_addr,
-												  GDB_REGNO_S0);
-					next_index = (next_read_addr - address) / increment;
-				}
-				if (result != ERROR_OK) {
-					riscv_batch_free(batch);
-					goto error;
-				}
+	/* First read has just triggered. Result is in s1.
+	 * dm_data registers contain the previos value of s1 (garbage).
+	 */
+	if (dmi_write(target, DM_ABSTRACTAUTO,
+				set_field(0, DM_ABSTRACTAUTO_AUTOEXECDATA, 1)) != ERROR_OK)
+		return ERROR_FAIL;
 
-				uint64_t value64 = (((uint64_t)dmi_data1) << 32) | dmi_data0;
-				buf_set_u64(buffer + (next_index - 2) * size, 0, 8 * size, value64);
-				log_memory_access64(address + (next_index - 2) * size, value64, size, true);
+	/* Read garbage from dm_data0, which triggers another execution of the
+	 * program. Now dm_data contains the first good result (from s1),
+	 * and s1 the next memory value.
+	 */
+	if (dmi_read_exec(target, NULL, DM_DATA0) != ERROR_OK) {
+		dmi_write(target, DM_ABSTRACTAUTO, 0);
+		return ERROR_FAIL;
+	}
 
-				/* Restore the command, and execute it.
-				 * Now DM_DATA0 contains the next value just as it would if no
-				 * error had occurred. */
-				dmi_write_exec(target, DM_COMMAND, command, true);
-				next_index++;
+	uint32_t abstractcs;
 
-				dmi_write(target, DM_ABSTRACTAUTO,
-						1 << DM_ABSTRACTAUTO_AUTOEXECDATA_OFFSET);
+	if (wait_for_idle(target, &abstractcs))
+		return ERROR_FAIL;
 
-				ignore_last = 1;
+	info->cmderr = get_field(abstractcs, DM_ABSTRACTCS_CMDERR);
+	switch (info->cmderr) {
+	case CMDERR_NONE:
+		break;
+	default:
+		/*TODO: Here one could attempt to try again on CMDERR_BUSY */
```

Comment:
What happens if we encounter BUSY? It should not be fatal, but must result in a retry. That doesn't have to be right here. In the comment can you mention where it's handled?

Reply:
Handling this particular `CMDERR_BUSY` is quite complicated. If one retries the access, it is possible to get stuck in an infinite loop if the hardware hangs and always returns busy. Since, at this point, no element has been read from memory.

This one `CMDERR_BUSY` is rare. I've never encountered it during testing. It would be much easier to address it when there is a reproduction of this exact `CMDERR_BUSY`.

I can file an issue.

Reply:
OK. In that case please update the comment to say that (for now) we're not retrying on this busy condition.

Reply:
I don't think CMDERR_BUSY can really occur here. execute_abstract_command() above already checks abstractcs.cmderr, and if it's non-zero then returns ERROR_FAIL. So there is no way that we see CMDERR_BUSY when you execute the abstract command in response to the DMI read, because we're certain that the previous abstract command has already completed.